src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -0,0 +1,302 @@
+    }
+  }
+
+  /** functions against pipes **/
+
+  // the root function starts the pipeline at some root entity type in the workspace
+  private def entityRootFunc(context: SlickExpressionContext): PipeType = {
+    for {
+      entity <- entityQuery.findEntityByName(context.workspaceContext.workspaceId, context.rootType, context.rootName)
+    } yield (entity, 0)
+  }
+
+  // final func that gets an attribute off a workspace
+  private def workspaceAttributeRootFunc(attributeName: String)(context: SlickExpressionContext, graphPipeline: Option[PipeType]): ReadAction[Iterable[Attribute]] = {
+    (for {
+      workspace <- workspaceQuery.findByIdQuery(context.workspaceContext.workspaceId)
+      workspaceAttribute <- workspaceAttributeQuery if workspaceAttribute.workspaceId === workspace.id
+      attribute <- attributeQuery if workspaceAttribute.attributeId === attribute.id && attribute.name === attributeName
+    } yield attribute).result.map(x => Seq(attributeQuery.unmarshalAttributes(x.map((_, None))).getOrElse(attributeName, AttributeNull)))
+  }
+
+  // root func that gets an entity reference off a workspace
+  private def workspaceEntityRefRootFunc(entityRefName: String)(context: SlickExpressionContext): PipeType = {
+    for {
+      workspace <- workspaceQuery.findByIdQuery(context.workspaceContext.workspaceId)
+      workspaceAttribute <- workspaceAttributeQuery if workspaceAttribute.workspaceId === workspace.id
+      attribute <- attributeQuery if workspaceAttribute.attributeId === attribute.id && attribute.name === entityRefName
+      nextEntity <- entityQuery if attribute.valueEntityRef === nextEntity.id
+    } yield (nextEntity, attribute.listIndex.getOrElse(0))
+  }
+
+  // add pipe to an entity referenced by the current entity
+  private def entityNameAttributePipeFunc(entityRefName: String)(context: SlickExpressionContext, queryPipeline: PipeType): PipeType = {
+      for {
+      (entity, listIndex) <- queryPipeline
+      entityAttribute <- entityAttributeQuery if entity.id === entityAttribute.entityId
+      attribute <- attributeQuery if entityAttribute.attributeId === attribute.id && attribute.name === entityRefName
+      nextEntity <- entityQuery if attribute.valueEntityRef === nextEntity.id
+    } yield (nextEntity, listIndex * 10 + attribute.listIndex.getOrElse(0))
+  }
+
+  // filter attributes to only the given attributeName and convert to attribute
+  private def entityAttributeFinalFunc(attributeName: String)(context: SlickExpressionContext, queryPipeline: Option[PipeType]): ReadAction[Iterable[Attribute]] = {
+    // attributeForNameQuery will only contain attribute records of the given name but for possibly more than 1 entity
+    // and in the case of a list there will be more than one attribute record for an entity
+    val attributeForNameQuery = (for {
+      (entity, listIndex) <- queryPipeline.get
```

Comment:
in this case it is an error signifying a bug, not something the user should be able to do